learn/testing-ballerina-code/mocking.md

```diff
@@ -0,0 +1,418 @@
+---
+layout: ballerina-left-nav-pages
+title: Mocking
+description: Learn how to use Ballerina's built-in mocking API provided by the test to test modules
+ independent from other modules and external endpoints.
+keywords: ballerina, programming language, testing, mocking
+permalink: /learn/testing-ballerina-code/mocking/
+active: mocking
+redirect_from:
+  - /v1-2/learn/testing-ballerina-code/mocking
+  - /learn/testing-ballerina-code/mocking
+---
+
+# Mocking
+
+Mocking is useful to control the behavior of functions and objects to control the communication with other modules and external endpoints. A mock can be created by defining return values or replacing the entire object or function with a user-defined equivalent. This feature will help you to test the Ballerina code independently from other modules and external endpoints.
+
+
+## Mocking Objects
+
+The test module provides capabilities to mock an object for unit testing. This allows you to control the behavior of object member functions and values of member fields via stubbing or replacing the entire object with a user-defined equivalent. This feature will help you to test the Ballerina code independently from other modules and external endpoints.
+
+Mocking of objects can be done in two ways.
+
+1. Creating a test double (providing an equivalent object in place of the real)
+2. Stubbing the member function or member variable (specifying the behavior of functions and values of variables)
+
+
+### Creating a test double
+
+You can write a custom mock object and substitute it in place of the real object. The custom object should be made structurally equivalent to the real object via the mocking features in the test module.
+
+***Example:***
+
+Let's make changes to the example in the [Quick Start](/learn/testing-ballerina-code/testing-quick-start) to define a test
+ double for the  `clientEndpont` object.
+
+Note that only the `get` function is implemented since it is the only function used in the sample. Attempting to call
```

Comment:
```suggestion
>**Note:** Only the `get` function is implemented since it is the only function used in the sample. Attempting to call
```
Add the formatting of a note.